SpatialGDK/Source/SpatialGDK/Private/Utils/ActorGroupManager.cpp

```diff
@@ -0,0 +1,202 @@
+	return false;
+}
+
+void UActorGroupManager::ValidateOffloadingSettings(TMap<FName, FActorClassSet> OldActorGroups, TMap<FName, FActorClassSet>* ActorGroups,
+	TSet<FName> OldWorkerTypes, TSet<FName>* WorkerTypes, FWorkerAssociation& WorkerAssociation)
+{
+	if (ActorGroups->Num() == 0)
+	{
+		ActorGroups->Append(UActorGroupManager::DefaultActorGroups());
+	}
+
+	// Check for empty Worker Types
+	if (WorkerTypes->Num() == 0)
+	{
+		WorkerTypes->Append(UActorGroupManager::DefaultWorkerTypes());
+	}
+
+	TArray<FName> ActorGroupKeys;
+	ActorGroups->GetKeys(ActorGroupKeys);
+
+	TArray<FName> OldActorGroupKeys;
+	OldActorGroups.GetKeys(OldActorGroupKeys);
+
+	// Check for renamed Actor Group
+	if (ActorGroups->Num() == OldActorGroups.Num())
+	{
+		FName FromActorGroup, ToActorGroup;
+		if (GetFirstDifferentValue(OldActorGroupKeys, ActorGroupKeys, FromActorGroup)
+			&& GetFirstDifferentValue(ActorGroupKeys, OldActorGroupKeys, ToActorGroup))
+		{
+			if (WorkerAssociation.ActorGroupToWorker.Contains(FromActorGroup))
+			{
+				WorkerAssociation.ActorGroupToWorker.Add(ToActorGroup, WorkerAssociation.ActorGroupToWorker.FindRef(FromActorGroup));
+				WorkerAssociation.ActorGroupToWorker.Remove(FromActorGroup);
+			}
+		}
+	}
+
+	// Check for invalid WorkerType (Currently just UnrealClient)
+	if (WorkerTypes->Contains(FName(*SpatialConstants::ClientWorkerType)))
+	{
+		WorkerTypes->Remove(FName(*SpatialConstants::ClientWorkerType));
+	}
+
+	// Check for renamed WorkerType
+	if (WorkerTypes->Num() == OldWorkerTypes.Num())
+	{
+		FName FromWorkerType, ToWorkerType;
+		if (GetFirstDifferentValue(OldWorkerTypes.Array(), WorkerTypes->Array(), FromWorkerType) &&
+			GetFirstDifferentValue(WorkerTypes->Array(), OldWorkerTypes.Array(), ToWorkerType))
+		{
+			for (auto Entry = WorkerAssociation.ActorGroupToWorker.CreateConstIterator(); Entry; ++Entry)
+			{
+				if (Entry->Value == FromWorkerType)
+				{
+					WorkerAssociation.ActorGroupToWorker.Add(Entry->Key, ToWorkerType);
+				}
+			}
+		}
+	}
+
+	// Remove any keys for deleted actor groups.
+	TArray<FName> Keys;
+	WorkerAssociation.ActorGroupToWorker.GetKeys(Keys);
+	for (FName Key : Keys)
+	{
+		if (!ActorGroupKeys.Contains(Key))
+		{
+			WorkerAssociation.ActorGroupToWorker.Remove(Key);
+		}
+	}
+
+	FName FirstWorkerType = WorkerTypes->Array()[0];
+
+	// Add default key for any new actor groups.
+	for (FName ActorGroup : ActorGroupKeys)
+	{
+		if (!WorkerAssociation.ActorGroupToWorker.Contains(ActorGroup))
+		{
+			WorkerAssociation.ActorGroupToWorker.Add(ActorGroup, FirstWorkerType);
+		}
+	}
+
+	// Replace any now invalid Worker Types with FirstWorkerType.
+	for (auto Entry = WorkerAssociation.ActorGroupToWorker.CreateConstIterator(); Entry; ++Entry)
+	{
+		if (!WorkerTypes->Contains(Entry->Value))
+		{
+			WorkerAssociation.ActorGroupToWorker.Add(Entry->Key, FirstWorkerType);
+		}
+	}
+
+	// Check Actor base class is referenced somewhere, add to first Actor Group if not.
+	bool bFoundBaseActor = false;
```

Comment:
This is a bit fragile because users are still able to nuke the whole Default actor group. I propose making the Default group read-only (at least in UI) to make it clear that there is always a fallback. Ideally, `GetActorGroupForClass` should return NAME_None if and only if the given class is not part of the `AActor` class hierarchy regardless of what the `WorkerAssociation` settings are.

Reply:
I also managed to crash my editor after deleting this group, pressing Ctrl+Z a few times and then trying to delete a worker type which appeared to be empty. Will try to repro it again and record the exact steps.

Reply:
- Add a new worker type and leave it as None
- Delete the one above it
- Delete the None text in the item you added and press Enter (the UI will revert back to None)
- Press Ctrl+Z - this will cause the item to become empty
- Delete the empty item - this causes the editor to crash

Reply:
I like having the default actor group as read only at first glance